src/backend/cuda/nvrtc/cache.cpp

```diff
@@ -0,0 +1,383 @@
+    CASE_STMT(af_bitshiftr_t);
+
+    CASE_STMT(af_min_t);
+    CASE_STMT(af_max_t);
+    CASE_STMT(af_cplx2_t);
+    CASE_STMT(af_atan2_t);
+    CASE_STMT(af_pow_t);
+    CASE_STMT(af_hypot_t);
+
+    CASE_STMT(af_sin_t);
+    CASE_STMT(af_cos_t);
+    CASE_STMT(af_tan_t);
+    CASE_STMT(af_asin_t);
+    CASE_STMT(af_acos_t);
+    CASE_STMT(af_atan_t);
+
+    CASE_STMT(af_sinh_t);
+    CASE_STMT(af_cosh_t);
+    CASE_STMT(af_tanh_t);
+    CASE_STMT(af_asinh_t);
+    CASE_STMT(af_acosh_t);
+    CASE_STMT(af_atanh_t);
+
+    CASE_STMT(af_exp_t);
+    CASE_STMT(af_expm1_t);
+    CASE_STMT(af_erf_t);
+    CASE_STMT(af_erfc_t);
+
+    CASE_STMT(af_log_t);
+    CASE_STMT(af_log10_t);
+    CASE_STMT(af_log1p_t);
+    CASE_STMT(af_log2_t);
+
+    CASE_STMT(af_sqrt_t);
+    CASE_STMT(af_cbrt_t);
+
+    CASE_STMT(af_abs_t);
+    CASE_STMT(af_cast_t);
+    CASE_STMT(af_cplx_t);
+    CASE_STMT(af_real_t);
+    CASE_STMT(af_imag_t);
+    CASE_STMT(af_conj_t);
+
+    CASE_STMT(af_floor_t);
+    CASE_STMT(af_ceil_t);
+    CASE_STMT(af_round_t);
+    CASE_STMT(af_trunc_t);
+    CASE_STMT(af_signbit_t);
+
+    CASE_STMT(af_rem_t);
+    CASE_STMT(af_mod_t);
+
+    CASE_STMT(af_tgamma_t);
+    CASE_STMT(af_lgamma_t);
+
+    CASE_STMT(af_notzero_t);
+
+    CASE_STMT(af_iszero_t);
+    CASE_STMT(af_isinf_t);
+    CASE_STMT(af_isnan_t);
+
+    CASE_STMT(af_sigmoid_t);
+
+    CASE_STMT(af_noop_t);
+
+    CASE_STMT(af_select_t);
+    CASE_STMT(af_not_select_t);
+    }
+#undef CASE_STMT
+}
+
+template<typename T>
+std::string toString(T value) {
+    return std::to_string(value);
+}
+
+template std::string toString<int               >(int               );
+template std::string toString<long              >(long              );
+template std::string toString<long long         >(long long         );
+template std::string toString<unsigned          >(unsigned          );
+template std::string toString<unsigned long     >(unsigned long     );
+template std::string toString<unsigned long long>(unsigned long long);
+template std::string toString<float             >(float             );
+template std::string toString<double            >(double            );
+template std::string toString<long double       >(long double       );
+
+
+template<> std::string toString(bool val) {
+    return std::string(val ? "true" : "false");
+}
+
+template<> std::string toString(af_op_t val) {
+    return getOpEnumStr(val);
+}
+
+template<> std::string toString(const char* str) {
+    return std::string(str);
+}
+
+Kernel getKernel(const string& nameExpr, const string& source,
+                 const vector< TemplateArg >& targs,
+                 const vector< string >& compileOpts) {
+    vector< string > args;
+    args.reserve(targs.size());
+
+    std::transform(targs.begin(), targs.end(), std::back_inserter(args),
+            [](const TemplateArg& arg) -> std::string {
+                 return arg._tparam;
+            });
+
+    std::string tInstance = nameExpr + "<" + args[0];
```

Comment:
Nit: This code block could be a function.

Reply:
It isn't being used anywhere else even in the very same source file.

Reply:
I was saying this to better document what is going on without reading the code. Its not a big deal.